tests/examples/testing-unit-java/src/test/java/com/pulumi/example/unittest/WebserverInfraTest.java

```diff
@@ -0,0 +1,82 @@
+package com.pulumi.example.unittest;
+
+import com.pulumi.aws.ec2.Instance;
+import com.pulumi.aws.ec2.SecurityGroup;
+import com.pulumi.aws.ec2.outputs.SecurityGroupIngress;
+import com.pulumi.test.Mocks;
+import com.pulumi.test.PulumiTest;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.TestFactory;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import static com.pulumi.test.PulumiTest.extractValue;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatPredicate;
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
+
+class WebserverInfraTest {
+
+    @AfterAll
+    static void cleanup() {
+        PulumiTest.cleanup();
+    }
+
+    @TestFactory
+    Stream<DynamicTest> testInstanceHasNameTag() {
+        var result = PulumiTest
+                .withMocks(new MyMocks())
+                .runTest(WebserverInfra::stack);
+
+        var instance = result.resources().stream()
+                .filter(r -> r instanceof Instance)
+                .map(r -> (Instance) r)
+                .findFirst();
+        assertThat(instance).isPresent();
+
+        return Stream.of(
+                dynamicTest("instance has name tag", () -> {
```

Comment:
I've used dynamic tests to be able to call run only once, I can easily rewrite this to be invoked many times like in the C# test if we don't want to use advanced features of JUnit 5 for some reason.

IMO this makes the test cleaner but I'm biased ;)

Reply:
I don't know. We should try to have the least surprising code for the Java user, I don't know what that would be. If this is idiomatic JUnit then OK. The idea is to be as much familiar to the people reading this as possible; there's a lot of Pulumi specific content to absorb and we want to be as simple as possible in every other aspect.

Reply:
the simplest but less clean approach would be to call `run` multiple times like c# does, but JUnit 5 gives this advanced ability to use dynamic tests

Reply:
Slight perference for the simplest apprioach but OK to leave as is since you say this is standard JUnit 5.